sheets/api.py

```diff
@@ -310,6 +312,91 @@ def batch_update_sheet_cells(self, sheet_id, request_objects):
         )
 
 
+def build_drive_service(credentials=None):
+    """
+    Builds the Google API client Drive service for API functionality that cannot be implemented correctly
+    with pygsheets.
+
+    Args:
+        credentials (google.oauth2.credentials.Credentials or None): Credentials to be used by the
+            Google Drive client
+
+    Returns:
+        googleapiclient.discovery.Resource: The Drive API service. The methods available on this resource are
+            defined dynamically (ref: http://googleapis.github.io/google-api-python-client/docs/dyn/drive_v3.html)
+    """
+    credentials = credentials or get_credentials()
+    return build("drive", "v3", credentials=credentials, cache_discovery=False)
+
+
+def batch_share_callback(
+    request_id, response, exception
+):  # pylint: disable=unused-argument
+    """
+    A callback function given to the Google API client's new_batch_http_request(). Called for the result
+    of each individual request executed in a batch API call.
+    Ref: https://developers.google.com/drive/api/v3/batch#using-client-libraries
+    NOTE: If *any* of the individual share requests in the batch fail, this callback is called and passed an exception
+    for *all* of the individual requests whether they contained a bad email address or not. There is no way to
+    determine the specific failed emails from the arguments to this function.
+
+    Args:
+        request_id (str): The id of the individual batch request
+        response (dict): The API response body if no error occurred
+        exception (googleapiclient.errors.HttpError): The error, if one occurred
+    """
+    if exception:
+        raise FailedBatchRequestException from exception
+
+
+def share_drive_file_with_emails(file_id, emails_to_share, credentials=None):
+    """
+    Shares a Drive file with multiple recipients. First attempts to share with the list of emails via a batch request.
+    If that fails, attempts to share with each email address individually. If a batch share request fails, the file is
+    not shared with any of the emails, and there is no way to get reliable information about which specific emails
+    failed. This function defaults to individual sharing so valid emails will still be shared to if the batch request
+    fails.
+
+    Args:
+        file_id (str): The Drive file id
+        emails_to_share (list of str): Email addresses that will be added as shared users for the given file
```

Comment:
If this is an empty list, as it seems to be in the default in settings.py, what should we do here? Does the Google API error if you ask to set permissions when there are zero users specified?